### PR TITLE
New: Treepicker

### PR DESCRIPTION
**Pull request by camjc**
- Stateful treepicker component.
- Fix overflow when there are a lot of nodes.
- Make tabs work.
- Ensure ids are strings throughout.
- Finish node expanders.
- Enable expander on a per node basis.
- Still need to put in a modal and wire up an Apply button.

https://trello.com/c/o4SSEkPa/4836-geo-picker-tests-3d

screen shot 2016-01-12 at 4 45 39 pm

### CODE REVIEW

File: src/components/adslotUi/TreePickerComponent.js [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+
+    newSelected[rootTypeId].push(node);
+    this.setState({ selectedNodesByRootType: newSelected });
+  }
+
+  removeNode(node) {
+    const newSelected = this.state.selectedNodesByRootType;
+    const { rootTypeId } = node;
+    newSelected[rootTypeId] = _.reject(newSelected[rootTypeId], { id: node.id });
+    if (_.isEmpty(newSelected[rootTypeId])) {delete newSelected[this.state.rootType.id];}
```
Comment (omgaz) [inline src/components/adslotUi/TreePickerComponent.js]: spaces around `{}`

File: test/components/adslotUi/TreePickerPureComponentTest.js [hunk 2]
```diff
@@ -48,8 +48,11 @@ describe('TreePickerPureComponent', () => {
         emptyIcon: 'http://placehold.it/70x70',
         isRequired: true,
       },
+      { label: 'Audiences', id: 1, icon: 'http://placehold.it/16x16', isRequired: false },
```
Comment (omgaz) [inline test/components/adslotUi/TreePickerPureComponentTest.js]: string everything

File: src/styles/adslotUi/TreePickerSelected.scss [hunk 3]
```diff
@@ -1,14 +1,30 @@
...
     margin: 0;
   }
 
+  &-scrollable {
+    $scrollable-max-height: 410px + $header-height;
+    max-height: $scrollable-max-height;
+    overflow: auto;
+
+    &.is-short {
+      max-height: $scrollable-max-height - 40px;
```
Comment (omgaz) [inline src/styles/adslotUi/TreePickerSelected.scss]: Var

File: src/styles/adslotUi/TreePickerNode.scss [hunk 4]
```diff
@@ -1,7 +1,20 @@
...
+@import '../icon';
 
 .treepickernode-component {
 
+  &:hover {
+    .treepickernode-component-expander {
+      background-image: $icon-expand;
+      cursor: pointer;
+      display: inline-block;
+      height: 16px;
```
Comment (omgaz) [inline src/styles/adslotUi/TreePickerNode.scss]: Maybe have a var inside icon called size that we can use for this and width

File: src/components/adslotUi/TreePickerComponent.js (hunk 1 above)
Comment (camjc) [inline src/components/adslotUi/TreePickerComponent.js]: its an `if` block not an `Object` block. Not supposed to have spaces.

File: src/styles/adslotUi/TreePickerNode.scss (hunk 4 above)
Comment (camjc) [inline src/styles/adslotUi/TreePickerNode.scss]: I agree but we're going to refactor this within the next fortnight anyway so I'm happy to leave this if you are.

Comment (omgaz) [inline src/styles/adslotUi/TreePickerNode.scss]: Yea, good point

**Comment by vinteo** [comment]
LGTM

File: src/components/adslotUi/TreePickerSelectedComponent.js [hunk 5]
```diff
@@ -50,35 +50,36 @@ const TreePickerSelectedComponent = ({
...
-
-          <GridRow type="subfooter">
-            <GridCell stretch />
-            <GridCell>Average</GridCell>
-            <GridCell>{valueFormatter(averagesByRootType[rootTypeId])}</GridCell>
-          </GridRow>
-
-        </Grid>
-      )}
+      <div className={`treepickerselected-component-scrollable${unresolvedRootTypes ? ' is-short' : ''}`}>
```
Comment (omgaz) [inline src/components/adslotUi/TreePickerSelectedComponent.js]: https://github.com/JedWatson/classnames#usage-with-reactjs

**Comment by vinteo** [comment]
Merging...

**Pull request merged**